### PR TITLE
Some miscellaneous builder improvements

### PR DESCRIPTION
**Pull request by sampsyo**
Just a few holes addressed on the journey to #1240. Not much exciting to see here. I also happened across a mistake in a fud error message and fixed that too.

### CODE REVIEW

**Comment by rachitnigam** [comment]
Ah, I ended up making some of these changes in #1295 as well. Also, the interpreter does not support sequential memories so importing it default might be breaking things? In general, I’d like to revisit the decision to have a default import list include at all

File: calyx-py/calyx/builder.py [hunk 1]
```diff
@@ -404,24 +432,21 @@ def infer_width(expr):
...
-    elif (
-        prim == "std_mult_pipe"
-        or prim == "std_smult_pipe"
-        or prim == "std_mod_pipe"
-        or prim == "std_smod_pipe"
-        or prim == "std_div_pipe"
-        or prim == "std_sdiv_pipe"
-    ):
+        if prim == "seq_mem_d1":
+            if port_name == "read_en":
```
Comment (cgyurgyik) [inline calyx-py/calyx/builder.py:446]:
```suggestion
        if prim == "seq_mem_d1" and port_name == "read_en":
```

**Comment by sampsyo** [comment]
> Ah, I ended up making some of these changes in https://github.com/cucapra/calyx/pull/1295 as well.

Ah, nice—sorry for the race condition!! I'll let that one go first, and then fix this one up to adapt.

> Also, the interpreter does not support sequential memories so importing it default might be breaking things? In general, I’d like to revisit the decision to have a default import list include at all

Good idea. Maybe we can implicitly import the first time you use a thing. Or maybe we want to offer some way to explicitly access things from the various libraries. Or maybe just the simplest/dumbest thing, which is that you have to explicitly ask to build imports (that's probably the right way).

File: calyx-py/calyx/builder.py (hunk 1 above)
Comment (sampsyo) [reply calyx-py/calyx/builder.py:446]: I was thinking this and then skipped over the simplification to emphasize the outer case, which adds a different set of ports (which happens to be a set of size 1)… hope that's cool!

**Comment by sampsyo** [comment]
I added a new explicit function `prog.import_` for adding `import` statements. That function also avoids duplicate imports, so it made it easy to implicitly import the `binary_operator` library when using `std_add` the first time. Of course there could be a much better system for importing libraries, but hopefully this at least avoids the proximate problem of importing the `memories` library when it's unused.

File: calyx-py/calyx/builder.py [hunk 2]
```diff
@@ -99,8 +99,14 @@ def comb_group(self, name: str) -> "GroupBuilder":
         return builder
 
     def cell(
-        self, name: str, comp: ast.CompInst, is_external=False, is_ref=False
+        self, name: str, comp: Union[ast.CompInst, "ComponentBuilder"],
+        is_external=False, is_ref=False
     ) -> "CellBuilder":
```
Comment (rachitnigam) [inline calyx-py/calyx/builder.py:111]: How can this feature be used? Or maybe, how does it show up in code?

Comment (sampsyo) [reply calyx-py/calyx/builder.py:111]: It's for instantiating components that you just recently built with the builder! Like this in AOC day 1:
https://github.com/cucapra/aoc2022-calyx/blob/e63cff9055fa7e183dabab6993f2dee26376e6ab/1/accelgen.py#L103

Comment (rachitnigam) [reply calyx-py/calyx/builder.py:111]: Awesome! Makes sense

**Comment by rachitnigam** [comment]
LGTM!

**Pull request merged**